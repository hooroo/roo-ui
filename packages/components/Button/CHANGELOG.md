# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="0.0.3"></a>
## [0.0.3](https://github.com/hooroo/roo-ui/compare/v0.0.2...v0.0.3) (2018-02-01)




**Note:** Version bump only for package @roo-ui/button

<a name="0.0.2"></a>
## [0.0.2](https://github.com/hooroo/roo-ui/compare/v0.0.1...v0.0.2) (2018-02-01)




**Note:** Version bump only for package @roo-ui/button

<a name="0.0.1"></a>
## 0.0.1 (2018-02-01)




**Note:** Version bump only for package @roo-ui/button
