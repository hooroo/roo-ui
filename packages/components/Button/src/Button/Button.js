import system from 'system-components';
import { css } from 'styled-components';
import { themeGet, complexStyle } from 'styled-system';
import { darken } from 'polished';

const background = (props) => {
  const styles = complexStyle({
    prop: 'buttonStyle',
    key: 'buttons',
  })(props);

  return props.bg
    ? themeGet(`colors.${props.bg}`, props.bg)(props)
    : styles.backgroundColor;
};

const hoverBackground = props => darken(0.1, background(props));

const Button = system({
  is: 'button',
  buttonStyle: 'default',
  boxShadow: null,
  color: null,
  bg: null,
  border: 2,
  borderColor: 'transparent',
  borderRadius: 'default',
  fontSize: 'sm',
  fontWeight: 'bold',
  letterSpacing: 'wide',
  lineHeight: 'normal',
  px: 5,
  py: 3,
  m: 0,
});

export default Button.extend`
  font-family: ${themeGet('fontFamily')};
  transition: ${themeGet('transitions.default')};
  display: inline-block;
  cursor: pointer;
  outline: 0;
  appearance: none;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;

  &:focus {
    box-shadow: ${themeGet('shadows.focus')};
  }

  &:hover {
    background-color: ${hoverBackground};
  }

  &[disabled] {
    opacity: 0.7;
    cursor: not-allowed;
  }

  ${props => props.rounded && css`
    border-radius: ${themeGet('radii.rounded')};
  `}

  ${props => props.outlined && css`
    background-color: transparent;
    color: ${background};
    border-color: ${background};

    &:hover {
      background-color: transparent;
      color: ${hoverBackground};
      border-color: ${hoverBackground};
    }
  `}
`;
