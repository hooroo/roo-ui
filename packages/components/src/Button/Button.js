import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';
import { darken } from 'polished';
import tag from 'clean-tag';
import { themeGet, space, color, boxShadow, buttonStyle } from 'styled-system';

const getBackground = props =>
  color(props).backgroundColor || buttonStyle(props).backgroundColor;

const Button = styled(tag.button)`
  display: inline-block;
  margin: 0;
  padding: ${themeGet('space.3')} ${themeGet('space.5')};
  font-size: ${themeGet('fontSizes.sm')};
  font-weight: ${themeGet('fontWeights.bold')};
  letter-spacing: ${themeGet('letterSpacings.wide')};
  line-height: ${themeGet('lineHeights.normal')};
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  border: ${themeGet('borders.2')};
  border-color: transparent;
  border-radius: ${themeGet('radii.default')};
  outline: 0;
  transition: ${themeGet('transitions.default')};
  cursor: pointer;
  appearance: none;

  ${buttonStyle}
  ${space}
  ${color}
  ${boxShadow}

  &:hover:not(:disabled) {
    background-color: ${props => darken(0.1, getBackground(props))};
  }

  &:focus {
    box-shadow: ${themeGet('shadows.focus')};
  }

  &:disabled {
    opacity: 0.7;
    cursor: not-allowed;
  }

  ${props => props.rounded && css`
    border-radius: ${themeGet('radii.rounded')};
  `}
`;

Button.propTypes = {
  ...buttonStyle.propTypes,
  ...space.propTypes,
  ...color.propTypes,
  ...boxShadow.propTypes,
  primary: PropTypes.bool,
  rounded: PropTypes.bool,
};

Button.defaultProps = {
  buttonStyle: 'default',
  blacklist: Object.keys(Button.propTypes),
};

export default Button;
