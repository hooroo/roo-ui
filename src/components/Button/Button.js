import styled from '@emotion/styled';
import { css } from '@emotion/core';
import PropTypes from 'prop-types';
import { darken } from 'polished';
import {
  space,
  color,
  boxShadow,
  display,
  variant,
  width,
} from 'styled-system';
import { themeGet } from '@styled-system/theme-get';
import get from 'lodash/get';

const FALLBACK_BG_COLOR = '#000';

const buttonStyle = variant({ key: 'buttons' });

const getBackground = props =>
  get(color(props), 'backgroundColor') ||
  get(buttonStyle(props), 'backgroundColor') ||
  FALLBACK_BG_COLOR;

const Button = styled.button`
  margin: 0;
  padding: ${themeGet('space.3')} ${themeGet('space.6')};
  font-size: ${themeGet('fontSizes.base')};
  font-weight: ${themeGet('fontWeights.bold')};
  letter-spacing: ${themeGet('letterSpacings.wide')};
  line-height: ${themeGet('lineHeights.normal')};
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  border: ${themeGet('borders.2')};
  border-color: transparent;
  border-radius: ${themeGet('radii.default')};
  outline: 0;
  transition: background-color ${themeGet('transitions.default')};
  cursor: pointer;
  appearance: none;

  ${display} ${buttonStyle} ${space} ${color} ${boxShadow} ${width} &:hover {
    background-color: ${props => darken(0.1, getBackground(props))};
  }

  &:hover:disabled {
    background-color: ${props => getBackground(props)};
  }

  &:focus {
    box-shadow: ${themeGet('shadows.focus')};
  }

  &:disabled {
    opacity: ${themeGet('opacity.disabled')};
    cursor: not-allowed;
  }

  ${props =>
    props.rounded &&
    css`
      border-radius: ${themeGet('radii.rounded')(props)};
    `} ${props =>
  props.block &&
  css`
    width: 100%;
  `};
`;

Button.propTypes = {
  ...variant.propTypes,
  ...space.propTypes,
  ...color.propTypes,
  ...boxShadow.propTypes,
  rounded: PropTypes.bool,
  block: PropTypes.bool,
};

Button.defaultProps = {
  variant: 'default',
  display: 'inline-block',
};

export default Button;
