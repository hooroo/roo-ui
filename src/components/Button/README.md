# Button

<!-- STORY -->

## Installation

```shell
$ yarn add roo-ui
```

## Example

```js
import { Button } from 'roo-ui/components';

export default <Button>Hello world</Button>;
```

## Example with LoadingIndicator

You may wish to conditionally add loading state to a button in some scenarios (i.e. long running tasks initiated via a button click). This can be achieved via composition.  
For example:

```js
import { Button, LoadingIndicator } from 'roo-ui/components';

export default (
  <Button disabled={isLoading}>
    {isLoading ? <LoadingIndicator color="white" size={21} /> : 'Hello world'}
  </Button>
);
```

## Properties

| Name       | Description                | Type     | Default  | Required? |
| :--------- | :------------------------- | :------- | :------- | :-------- |
| `variant`  | use a theme variant        | `string` | 'default | -         |
| `rounded`  | use a rounded border       | `bool`   | false    | -         |
| `block`    | render a full width button | `bool`   | false    | -         |
| `disabled` | disable the button         | `bool`   | false    | -         |

## Customization

This component can be customized with [styled-system](https://jxnblk.com/styled-system) by passing props for [space](https://jxnblk.com/styled-system#space-theming), [color](http://jxnblk.com/styled-system/table#core), or [box shadow](http://jxnblk.com/styled-system/table#misc).
