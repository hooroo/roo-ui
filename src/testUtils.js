import React from 'react';
import { mount, shallow } from 'enzyme';

export const mountWithTheme = (tree, theme) => {
  const themedTree = React.cloneElement(tree, { theme });
  return mount(themedTree);
};

export const shallowWithTheme = (tree, theme) => {
  const themedTree = React.cloneElement(tree, { theme });
  return shallow(themedTree);
};
