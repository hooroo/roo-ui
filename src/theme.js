import { rem } from 'polished';
import range from 'lodash/range';

import * as icons from './defaultIcons';

const namedColors = {
  white: '#FFFFFF',
  brightSun: '#FBCB3B',
  red: '#E00E00',
  maroon: '#870000',
  blue: '#8DE2E0',
  lightBlue: '#E7F7F7',
  orange: '#ED710B',
  lightOrange: '#FCEBCD',
  green: '#35A509',
  lightGreen: '#DEF2DE',
};

const brand = {
  primary: '#E40000',
  secondary: '#8DE2E0',
  tertiary: '#D20000',
  quaternary: '#F9F3E9',
};

const greys = {
  charcoal: '#323232',
  steel: '#666666',
  alto: '#DADADA',
  porcelain: '#F4F5F6',
  dusty: '#999999',
  mineShaft: '#222222',
};

const ui = {
  error: namedColors.orange,
  errorBackground: namedColors.lightOrange,
  info: namedColors.blue,
  infoBackground: namedColors.lightBlue,
  link: namedColors.red,
  linkHover: namedColors.maroon,
  success: namedColors.green,
  successBackground: namedColors.lightGreen,
};

const colors = {
  brand,
  greys,
  ui,
  ...namedColors,
};

const alertStyles = {
  default: { bg: 'greys.porcelain' },
  error: {
    bg: 'ui.errorBackground',
    icon: {
      name: 'warning',
      color: 'ui.error',
    },
  },
  info: {
    bg: 'ui.infoBackground',
    icon: {
      name: 'info',
      color: 'greys.charcoal',
    },
  },
  success: {
    bg: 'ui.successBackground',
    icon: {
      name: 'checkCircle',
      color: 'ui.success',
    },
  },
};

const borders = [0, '1px solid', '2px solid', '3px solid'];

const breakpoints = ['40rem', '52rem', '64rem'];

const buttons = {
  default: {
    color: 'white',
    backgroundColor: 'greys.charcoal',
  },
  primary: {
    color: 'white',
    backgroundColor: 'brand.primary',
  },
};

const fontFamily = 'Ciutadella, sans-serif';

const fontSizes = {
  xs: rem('13px'),
  sm: rem('15px'),
  base: rem('16px'),
  lg: rem('18px'),
  xl: rem('22px'),
  '2xl': rem('26px'),
  '3xl': rem('36px'),
  '4xl': rem('46px'),
};

const fontWeights = {
  normal: 500,
  bold: 600,
};

const letterSpacings = {
  normal: '0',
  wide: '0.125em',
};

const lineHeights = {
  tight: 1,
  normal: 1.35,
  loose: 1.5,
};

const maxWidths = {
  default: rem('1280px'),
};

const mediaQueries = breakpoints.map(
  n => `@media screen and (min-width: ${n})`,
);

const opacity = {
  disabled: 0.7,
};

const radii = {
  none: 0,
  default: '4px',
  rounded: '1000px',
};

const shadows = {
  none: 0,
  default: '0 1px 1px 0 rgba(0, 0, 0, 0.1)',
  heavy: '0 2px 2px 0 rgba(0, 0, 0, 0.1)',
  focus: `0 0 2px 2px ${colors.greys.dusty};`,
};

const space = range(0, 129, 4).map(i => rem(`${i}px`));

const textStyles = {
  caps: {
    textTransform: 'uppercase',
    fontWeight: fontWeights.bold,
  },

  h1: {
    fontSize: fontSizes['3xl'],
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.normal,
  },
  h2: {
    fontSize: fontSizes['2xl'],
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.normal,
  },
  h3: {
    fontSize: fontSizes.xl,
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.loose,
  },
  h4: {
    fontSize: fontSizes.lg,
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.loose,
  },
  h5: {
    fontSize: fontSizes.base,
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.loose,
  },
  h6: {
    fontSize: fontSizes.sm,
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.loose,
  },
  label: {
    fontSize: fontSizes.sm,
    lineHeight: lineHeights.loose,
    fontWeight: fontWeights.bold,
    color: colors.greys.charcoal,
  },
  text: {
    fontSize: fontSizes.base,
    lineHeight: lineHeights.loose,
    fontWeight: fontWeights.normal,
  },
};

const transitions = {
  default: '200ms ease-in',
};

export default {
  alertStyles,
  borders,
  breakpoints,
  buttons,
  colors,
  fontFamily,
  fontSizes,
  fontWeights,
  icons,
  letterSpacings,
  lineHeights,
  maxWidths,
  mediaQueries,
  opacity,
  radii,
  shadows,
  space,
  textStyles,
  transitions,
};
