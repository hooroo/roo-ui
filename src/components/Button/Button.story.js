import React from 'react';
import { storiesOf } from '@storybook/react';
import { withDocs } from 'storybook-readme';
import { boolean, select } from '@storybook/addon-knobs/react';

import Button from './';
import README from './README.md';
import LoadingIndicator from '../LoadingIndicator';

storiesOf('Components|Button', module)
  .addDecorator(withDocs(README))
  .add('default', () => (
    <Button
      variant={select('Variant', ['default', 'primary'], 'default')}
      rounded={boolean('Rounded', false)}
      block={boolean('Block', false)}
      disabled={boolean('Disabled', false)}
    >
      Hello world
    </Button>
  ))
  .add('with loading indicator', () => {
    const isLoading = boolean('isLoading?', true);
    return (
      <Button
        variant={select('Variant', ['default', 'primary'], 'default')}
        rounded={boolean('Rounded', false)}
        block={boolean('Block', false)}
        disabled={isLoading}
      >
        {isLoading ? (
          <LoadingIndicator color="white" size={21} />
        ) : (
          'Hello world'
        )}
      </Button>
    );
  });
