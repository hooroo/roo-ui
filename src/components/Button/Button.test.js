import React from 'react';
import theme from 'theme';
import { shallowWithTheme } from 'testUtils';
import { axe } from 'jest-axe';

import Button from '.';

describe('<Button />', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowWithTheme(<Button>Hello world</Button>, theme);
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('has no accessibility errors', async () => {
    expect(await axe(wrapper.html())).toHaveNoViolations();
  });

  describe('primary', () => {
    beforeEach(() => {
      wrapper = shallowWithTheme(
        <Button variant="primary">Hello world</Button>,
        theme,
      );
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('rounded', () => {
    beforeEach(() => {
      wrapper = shallowWithTheme(<Button rounded>Hello world</Button>, theme);
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('block', () => {
    beforeEach(() => {
      wrapper = shallowWithTheme(<Button block>Hello world</Button>, theme);
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
