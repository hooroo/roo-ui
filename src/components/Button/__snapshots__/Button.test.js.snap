// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> block renders correctly 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`<Button /> primary renders correctly 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`<Button /> renders correctly 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`<Button /> rounded renders correctly 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
